custom/icds_reports/views.py

```diff
@@ -1729,3 +1731,45 @@ def get(self, request, *args, **kwargs):
             return export_response(sync.get_file_from_blobdb(), 'csv', zip_name)
         except NotFound:
             raise Http404
+
+
+class DishaAPIView(View):
+
+    def message(self, message_name):
+        state_names = ", ".join(self.valid_state_names)
+        error_messages = {
+            "missing_date": "Please specify valid month and year",
+            "invalid_month": "Please specify a month that's older than a month and 5 days",
+            "invalid_state": "Please specify one of {} as state_name".format(state_names),
+        }
+        return {"message": error_messages[message_name]}
+
+    @method_decorator([api_auth, toggles.ICDS_DISHA_API.required_decorator()])
+    def get(self, request, *args, **kwargs):
+        try:
+            month = int(request.GET.get('month'))
+            year = int(request.GET.get('year'))
+        except (ValueError, TypeError):
+            return JsonResponse(self.message('missing_date'), status=400)
+
+        # Can return only one month old data if today is after 5th, otherwise
+        #   can return two month's old data
+        query_month = date(year, month, 1)
+        today = date.today()
+        current_month = today - relativedelta(months=1) if today.day <= 5 else today
+        if query_month > current_month:
+            return JsonResponse(self.message('invalid_month'), status=400)
+
+        state_name = self.request.GET.get('state_name')
+        if state_name not in self.valid_state_names:
+            return JsonResponse(self.message('invalid_state'), status=400)
+
+        data = DishaDump(state_name, query_month).get_data()
+        if not data:
+            return JsonResponse({"message": "Data is not updated for this month"})
+        return HttpResponse(data, content_type='application/json')
+
+    @property
+    @quickcache([])
+    def valid_state_names(self):
+        return AwcLocationMonths.objects.values_list('state_name', flat=True).distinct()
```

Comment:
@sravfeyn this seems like a really inefficient way to get the list of state names. Does it have to come from submitted data? Could we not just get it straight from the location hierarchy?

At minimum you should switch to the `AwcLocation` table instead of `AwcLocationMonths`.

Reply:
Fixed it here https://github.com/dimagi/commcare-hq/pull/21901/commits/f32afe82299d2a0f0d33a8f2e2bac71443636e98

Unfortunately the name has to come from the external user.